File: MovieCatalogueV2/app/src/main/java/com/google/fdp/moviecataloguev2/networks/RetrofitClient.java
package com.google.fdp.moviecataloguev2.networks;

import android.util.Log;

import com.google.gson.FieldNamingPolicy;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by gama on 2019-07-26.
 * Addin Gama Bertaqwa
 * addingama@gmail.com
 */
public class RetrofitClient {
    public <S> S createService(Class<S> serviceClass) {
        Gson gson = new GsonBuilder().setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
                .setDateFormat("yyyy-MM-dd'T'HH:mm:ss")
                .create();
        Retrofit retrofit = new Retrofit.Builder().baseUrl("https://api.themoviedb.org")
                .addConverterFactory(GsonConverterFactory.create(gson))
                .client(getOkHttp().build())
                .build();

        return retrofit.create(serviceClass);
    }

    private OkHttpClient.Builder getOkHttp() {
        OkHttpClient.Builder builder = new OkHttpClient.Builder();
        HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor(new HttpLoggingInterceptor.Logger() {
            @Override
            public void log(String message) {
                Log.d("Movie Catalogue", message);
            }
        });

        interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);

        builder.addInterceptor(interceptor);



        builder.connectTimeout(360, TimeUnit.SECONDS);
        builder.readTimeout(360, TimeUnit.SECONDS);
        builder.writeTimeout(360, TimeUnit.SECONDS);

        return builder;
    }
}

File: MovieCatalog 3/app/src/main/java/com/alfianh/moviecatalog/network/RetrofitClient.kt
package com.alfianh.moviecatalog.network

import com.alfianh.moviecatalog.BuildConfig
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

object RetrofitClient {

  fun <S> createService(serviceClass: Class<S>): S {

    val httpClient = OkHttpClient.Builder()
    httpClient.addInterceptor { chain ->
      val original = chain.request()
      val originalHttpUrl = original.url()
      val url = originalHttpUrl.newBuilder().addQueryParameter("api_key", BuildConfig.API_KEY).build()
      val requestBuilder = original.newBuilder().url(url)
      val request = requestBuilder.build()
      chain.proceed(request)
    }

    return Retrofit.Builder()
        .baseUrl("https://api.themoviedb.org/3/")
        .addConverterFactory(GsonConverterFactory.create())
        .client(httpClient.build())
        .build()
        .create(serviceClass)
  }

}